### DISCUSSION TITLE
Modify a driver file out-of-tree

### DISCUSSION THREAD

**Discussion by mjm987** [zephyrproject-rtos/zephyr#49667] [Q&A]
I tried to patch a Zephyr driver file (eg. drivers/modem/gsm_ppp.c) out-of-tree by copying the driver file to the application src directory and modifying it...

But when building I get a "multiple definition of ...".
I tried to remove the original source in CMakeLists.txt by
list(FILTER Zephyr EXCLUDE REGEX ".*drivers/modem/gsm_ppp.c")
but it didn't work.

Same problem when cloning the driver as an EXTRA_MODULE.

PS: Possibly it might be done by cloning the driver as a EXTRA_MODULE and renaming all the CONFIG_* options, but this is definitively not what I want to do.

Any hints?

**Comment by aravinth-kalai** [comment]
Hi @mjm987 , 

Did you got any idea to fix this if yes please share your input,  I'm also having the same issue.

**Reply by ndrs-pst** [reply]
@aravinth-kalai, I think it may related to this discussion https://github.com/zephyrproject-rtos/zephyr/discussions/53210 and also here https://github.com/zephyrproject-rtos/zephyr/discussions/52229

**Reply by mjm987** [reply]
Unfortunately I didn't find a easy solution to copy/use the driver into the project directory so I created a local branch in the zephyr tree, patched the source there and rebase it every time I pull the zephyr master tree.

**Reply by ndrs-pst** [reply]
@mjm987, Absolutely, that's a great approach! 😄
A well-considered patch, if merged via a pull request into the mainline, can indeed help minimize your maintenance efforts.
Keep up the good work!
